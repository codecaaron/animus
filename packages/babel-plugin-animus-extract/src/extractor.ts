import type { NodePath, types as BabelTypes } from '@babel/core';
import type { CallExpression } from '@babel/types';

import { generateCSS } from './css-generator';
import { createRuntimeReplacement } from './transformer';
import type { AnimusExtractState, ComponentData, ExtractedChainCall } from './types';
import { generateComponentId } from './utils';

export function isAnimusBuilder(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  // Direct call: animus.styles()
  if (
    t.isMemberExpression(callee) &&
    t.isIdentifier(callee.object) &&
    (callee.object.name === 'animus' || callee.object.name === state.animusBinding)
  ) {
    return true;
  }

  // Chained call: animus.styles().states()
  if (t.isMemberExpression(callee) && t.isCallExpression(callee.object)) {
    const objectPath = path.get('callee.object') as NodePath<CallExpression>;
    return isAnimusBuilder(objectPath, state);
  }

  return false;
}

function isAnimusChainCall(path: NodePath<CallExpression>, state: AnimusExtractState): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  return t.isMemberExpression(callee) &&
    (isAnimusBuilder(path, state) ||
     (t.isCallExpression(callee.object) && isAnimusChainCall(path.get('callee.object') as NodePath<CallExpression>, state)));
}

export function extractAnimusComponent(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): void {
  const chain = extractChain(path, state);
  if (chain.length === 0) return;

  const componentData = processChain(chain, state);

  // Generate CSS for this component
  const css = generateCSS(componentData, state);
  if (css) {
    state.cssOutput.set(componentData.id, css);
  }

  // Store component data for JSX transformation
  if (componentData.name) {
    state.animusComponents.set(componentData.name, componentData);
  }

  // Replace the original expression with runtime version
  const replacement = createRuntimeReplacement(componentData, state);
  path.replaceWith(replacement);
}

function extractChain(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): ExtractedChainCall[] {
  const chain: ExtractedChainCall[] = [];
  let current: NodePath<CallExpression> | null = path;

  while (current && isAnimusChainCall(current, state)) {
    const { types: t } = state;
    const {callee} = current.node;

    if (t.isMemberExpression(callee) && t.isIdentifier(callee.property)) {
      const methodName = callee.property.name;
      const args = current.node.arguments;

      chain.unshift({ methodName, args });

      if (t.isCallExpression(callee.object)) {
        current = current.get('callee.object') as NodePath<CallExpression>;
      } else {
        current = null;
      }
    } else {
      current = null;
    }
  }

  return chain;
}

function processChain(
  chain: ExtractedChainCall[],
  state: AnimusExtractState
): ComponentData {
  const { types: t } = state;
  const componentData: ComponentData = {
    id: generateComponentId(),
  };

  for (const { methodName, args } of chain) {
    switch (methodName) {
      case 'styles':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.styles = extractObjectLiteral(args[0], state);
        }
        break;

      case 'states':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.states = extractObjectLiteral(args[0], state);
        }
        break;

      case 'variants':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.variants = extractObjectLiteral(args[0], state);
        }
        break;

      case 'groups':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.groups = extractObjectLiteral(args[0], state);
        }
        break;

      case 'props':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.props = extractObjectLiteral(args[0], state);
        }
        break;

      case 'asElement':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.element = args[0].value;
        }
        break;

      case 'displayName':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.name = args[0].value;
        }
        break;
    }
  }

  return componentData;
}

function extractObjectLiteral(
  node: BabelTypes.ObjectExpression,
  state: AnimusExtractState
): Record<string, any> {
  const { types: t } = state;
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) || t.isObjectMethod(prop)) {
      const key = getPropertyKey(prop, state);
      if (key && t.isObjectProperty(prop)) {
        const value = extractValue(prop.value, state);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

function getPropertyKey(
  prop: BabelTypes.ObjectProperty | BabelTypes.ObjectMethod,
  state: AnimusExtractState
): string | null {
  const { types: t } = state;

  if (t.isIdentifier(prop.key)) {
    return prop.key.name;
  } else if (t.isStringLiteral(prop.key)) {
    return prop.key.value;
  }

  return null;
}

function extractValue(node: BabelTypes.Node, state: AnimusExtractState): any {
  const { types: t } = state;

  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node, state);
  } else if (t.isArrayExpression(node)) {
    return node.elements.map(el => el ? extractValue(el, state) : null);
  } else if (t.isUnaryExpression(node) && node.operator === '-' && t.isNumericLiteral(node.argument)) {
    return -node.argument.value;
  }

  // For dynamic values, return undefined to skip extraction
  return undefined;
}
