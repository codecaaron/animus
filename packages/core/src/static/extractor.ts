import * as parser from '@babel/parser';
// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
import * as t from '@babel/types';

const traverse = (traverseDefault as any).default || traverseDefault;

import type { NodePath } from '@babel/traverse';

/**
 * Extracted style data from an Animus component
 */
export interface ExtractedStyles {
  componentName?: string;
  baseStyles?: Record<string, any>;
  variants?: Record<string, any> | Record<string, any>[];
  states?: Record<string, any>;
  groups?: string[];
  props?: Record<string, any>;
}

/**
 * Extract styles from Animus method chains in code
 */
export function extractStylesFromCode(code: string): ExtractedStyles[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const extractedComponents: ExtractedStyles[] = [];

  traverse(ast as any, {
    CallExpression(path: NodePath<t.CallExpression>) {
      // Look for .styles() calls
      if (
        t.isMemberExpression(path.node.callee) &&
        t.isIdentifier(path.node.callee.property, { name: 'styles' })
      ) {
        const extracted: ExtractedStyles = {};

        // Try to get component name from variable declaration
        const varDeclarator = path.findParent((p: NodePath) =>
          p.isVariableDeclarator()
        );
        if (varDeclarator && t.isVariableDeclarator(varDeclarator.node)) {
          if (t.isIdentifier(varDeclarator.node.id)) {
            extracted.componentName = varDeclarator.node.id.name;
          }
        }

        // Extract styles object from first argument
        const stylesArg = path.node.arguments[0];
        if (t.isObjectExpression(stylesArg)) {
          extracted.baseStyles = extractObjectLiteral(stylesArg);
        }

        // Look for chained method calls by traversing up the AST
        // We need to find the full chain starting from .styles()
        const chainedCalls: any[] = [];
        let currentPath: any = path;

        // Traverse up to find all chained calls
        while (currentPath) {
          if (
            t.isCallExpression(currentPath.node) &&
            t.isMemberExpression(currentPath.node.callee) &&
            t.isIdentifier(currentPath.node.callee.property)
          ) {
            chainedCalls.push({
              method: currentPath.node.callee.property.name,
              args: currentPath.node.arguments,
            });
          }

          // Check if parent is also a call expression (chained)
          if (
            currentPath.parentPath &&
            t.isMemberExpression(currentPath.parentPath.node) &&
            currentPath.parentPath.parentPath &&
            t.isCallExpression(currentPath.parentPath.parentPath.node)
          ) {
            currentPath = currentPath.parentPath.parentPath;
          } else {
            break;
          }
        }

        // Process chained calls in order
        const variantCalls: any[] = [];

        for (const call of chainedCalls) {
          if (call.method === 'variant' && call.args[0]) {
            const variantArg = call.args[0];
            if (t.isObjectExpression(variantArg)) {
              variantCalls.push(extractObjectLiteral(variantArg));
            }
          } else if (call.method === 'states' && call.args[0]) {
            const statesArg = call.args[0];
            if (t.isObjectExpression(statesArg)) {
              extracted.states = extractObjectLiteral(statesArg);
            }
          } else if (call.method === 'groups' && call.args[0]) {
            const groupsArg = call.args[0];
            if (t.isObjectExpression(groupsArg)) {
              const groups = extractObjectLiteral(groupsArg);
              extracted.groups = Object.keys(groups).filter(
                (key) => groups[key]
              );
            }
          } else if (call.method === 'props' && call.args[0]) {
            const propsArg = call.args[0];
            if (t.isObjectExpression(propsArg)) {
              extracted.props = extractObjectLiteral(propsArg);
            }
          }
        }

        // Store variants - if multiple, keep as array; if single, flatten
        if (variantCalls.length > 1) {
          extracted.variants = variantCalls;
        } else if (variantCalls.length === 1) {
          extracted.variants = variantCalls[0];
        }

        extractedComponents.push(extracted);
      }
    },
  });

  return extractedComponents;
}

/**
 * Extract a literal object from an AST ObjectExpression
 */
function extractObjectLiteral(node: t.ObjectExpression): Record<string, any> {
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) && !prop.computed) {
      const key = t.isIdentifier(prop.key)
        ? prop.key.name
        : t.isStringLiteral(prop.key)
          ? prop.key.value
          : null;

      if (key) {
        const value = extractValue(prop.value);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

/**
 * Extract a literal value from an AST node
 */
function extractValue(node: t.Node): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node);
  } else if (t.isArrayExpression(node)) {
    return node.elements
      .filter((el): el is t.Expression | t.SpreadElement => el !== null)
      .map((el) => (t.isExpression(el) ? extractValue(el) : undefined))
      .filter((v) => v !== undefined);
  } else if (t.isTemplateLiteral(node) && node.expressions.length === 0) {
    // Simple template literal with no expressions
    return node.quasis[0].value.cooked;
  }

  // Return undefined for non-literal expressions
  return undefined;
}
