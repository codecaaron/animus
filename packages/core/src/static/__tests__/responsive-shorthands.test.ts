import { extractStylesFromCode } from '../extractor';
import { CSSGenerator } from '../generator';
import { minimalSpace, testTheme } from './testConfig';

describe('Responsive Values and CSS Shorthands', () => {
  let generator: CSSGenerator;

  beforeEach(() => {
    generator = new CSSGenerator();
  });

  describe('CSS Shorthand Expansion', () => {
    it('expands margin shorthands correctly', () => {
      const code = `
        const Box = animus
          .styles({
            m: 0,
            mx: 10,
            my: 20,
            mt: 5,
            mb: 15,
            mr: 25,
            ml: 35,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('margin-shorthand-expansion');
    });

    it('expands padding shorthands correctly', () => {
      const code = `
        const Box = animus
          .styles({
            p: 10,
            px: 20,
            py: 30,
            pt: 15,
            pl: 25,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('padding-shorthand-expansion');
    });

    it('expands other shorthands (bg, size, area, inset)', () => {
      const code = `
        const Component = animus
          .styles({
            bg: 'red',
            size: 100,
            area: 'header',
            inset: 0,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('other-shorthand-expansion');
    });
  });

  describe('Responsive Array Syntax', () => {
    it('generates media queries for array syntax', () => {
      const code = `
        const Box = animus
          .styles({
            padding: [10, 20, 30],
            margin: [5, , 15], // sparse array
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('responsive-array-syntax');
    });

    it('handles responsive arrays with shorthands', () => {
      const code = `
        const Box = animus
          .styles({
            mx: [10, 20],
            py: [5, , 15],
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('responsive-array-shorthands');
    });
  });

  describe('Responsive Object Syntax', () => {
    it('generates media queries for object syntax', () => {
      const code = `
        const Box = animus
          .styles({
            padding: { _: 10, sm: 20, lg: 40 },
            margin: { _: 0, md: 'auto' },
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('responsive-object-syntax');
    });

    it('handles object syntax with shorthands', () => {
      const code = `
        const Box = animus
          .styles({
            mx: { _: 'auto', md: 20 },
            bg: { _: 'white', sm: 'gray', lg: 'black' },
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('responsive-object-shorthands');
    });
  });

  describe('Complex Nested Responsive Styles', () => {
    it('handles responsive values in pseudo-selectors', () => {
      const code = `
        const Button = animus
          .styles({
            padding: [10, 20],
            '&:hover': {
              bg: { _: 'gray', md: 'blue' },
              transform: 'scale(1.05)',
            },
            '&:active': {
              transform: { _: 'scale(0.95)', lg: 'scale(0.98)' },
            },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('responsive-nested-selectors');
    });
  });

  describe('Atomic Utilities with Shorthands', () => {
    it('generates atomic utilities for shorthand properties', () => {
      const code = `
        const Box = animus
          .styles({
            display: 'flex',
          })
          .groups({
            space: true,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      // Use only mx and py to demonstrate shorthand expansion
      const groupDefs = {
        space: {
          mx: minimalSpace.mx,
          py: minimalSpace.py,
        },
      };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        testTheme
      );
      expect(result.css).toMatchSnapshot('atomic-shorthand-utilities');
    });
  });
});
