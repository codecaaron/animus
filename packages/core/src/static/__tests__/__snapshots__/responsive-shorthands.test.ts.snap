// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Responsive Values and CSS Shorthands Atomic Utilities with Shorthands generates atomic utilities for shorthand properties: atomic-shorthand-utilities 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  display: flex;
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Responsive Values and CSS Shorthands CSS Shorthand Expansion expands margin shorthands correctly: margin-shorthand-expansion 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  margin: 0px;
  margin-left: 10px;
  margin-right: 10px;
  margin-top: 20px;
  margin-bottom: 20px;
  margin-top: 5px;
  margin-bottom: 15px;
  margin-right: 25px;
  margin-left: 35px;
}"
`;

exports[`Responsive Values and CSS Shorthands CSS Shorthand Expansion expands other shorthands (bg, size, area, inset): other-shorthand-expansion 1`] = `
"/* Base Styles */
.animus-Component-c9t {
  grid-area: header;
  background-color: red;
  width: 100px;
  height: 100px;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
}"
`;

exports[`Responsive Values and CSS Shorthands CSS Shorthand Expansion expands padding shorthands correctly: padding-shorthand-expansion 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  padding: 10px;
  padding-left: 20px;
  padding-right: 20px;
  padding-top: 30px;
  padding-bottom: 30px;
  padding-top: 15px;
  padding-left: 25px;
}"
`;

exports[`Responsive Values and CSS Shorthands Complex Nested Responsive Styles handles responsive values in pseudo-selectors: responsive-nested-selectors 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 10px;
}
.animus-Button-b6n:hover {
  background-color: gray;
  transform: scale(1.05);
}
@media screen and (min-width: 1024px) {
  .animus-Button-b6n:hover {
    background-color: blue;
  }
}
.animus-Button-b6n:active {
  transform: scale(0.95);
}
@media screen and (min-width: 1200px) {
  .animus-Button-b6n:active {
    transform: scale(0.98);
  }
}
@media screen and (min-width: 480px) {
  .animus-Button-b6n {
    padding: 20px;
  }
}"
`;

exports[`Responsive Values and CSS Shorthands Responsive Array Syntax generates media queries for array syntax: responsive-array-syntax 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  padding: 10px;
  margin: 5px;
}
@media screen and (min-width: 480px) {
  .animus-Box-b3x {
    padding: 20px;
    margin: 15px;
  }
}
@media screen and (min-width: 768px) {
  .animus-Box-b3x {
    padding: 30px;
  }
}"
`;

exports[`Responsive Values and CSS Shorthands Responsive Array Syntax handles responsive arrays with shorthands: responsive-array-shorthands 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  margin-left: 10px;
  margin-right: 10px;
  padding-top: 5px;
  padding-bottom: 5px;
}
@media screen and (min-width: 480px) {
  .animus-Box-b3x {
    margin-left: 20px;
    margin-right: 20px;
    padding-top: 15px;
    padding-bottom: 15px;
  }
}"
`;

exports[`Responsive Values and CSS Shorthands Responsive Object Syntax generates media queries for object syntax: responsive-object-syntax 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  padding: 10px;
  margin: 0px;
}
@media screen and (min-width: 768px) {
  .animus-Box-b3x {
    padding: 20px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-Box-b3x {
    padding: 40px;
  }
}
@media screen and (min-width: 1024px) {
  .animus-Box-b3x {
    margin: auto;
  }
}"
`;

exports[`Responsive Values and CSS Shorthands Responsive Object Syntax handles object syntax with shorthands: responsive-object-shorthands 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  margin-left: auto;
  margin-right: auto;
  background-color: white;
}
@media screen and (min-width: 1024px) {
  .animus-Box-b3x {
    margin-left: 20px;
    margin-right: 20px;
  }
}
@media screen and (min-width: 768px) {
  .animus-Box-b3x {
    background-color: gray;
  }
}
@media screen and (min-width: 1200px) {
  .animus-Box-b3x {
    background-color: black;
  }
}"
`;
